### ISSUE TITLE
Self-signed certificate and certificate authority throws exception

### ISSUE THREAD

**Issue by HarrPerson** [s3drive/app#53]
I use a self hosted (e.g. https://s3.local.lan) s3 storage (minio) with a self-signed certificate and certificate authority (CA). The CA is imported in android's certificate store within trusted credentials as "User trusted CA".

The self-signed CA works with local self hosted web pages, which suggests that the import works. For example it is possible to connect to s3 via browser (e.g. https://s3.local.lan) without CA warning or error.

When using s3drive, I receive the following error message / exception:

> _"HandshakeException: Handshake error in client (OS Error: CERTIFICATE_VERIFY_FAILED: unable to get local Issuer certificate (handshake.cc:393))"_

Any suggestions how to solve this issue?

**Comment by tomekit** [comment]
There is some work required on our end to make this work.

Initial phase likely won't recognize: "User trusted CA" added on the OS level, but instead would require loading e.g. `root.pem` (trusted CA) to the app itself.

Since this topic is getting more attention, we're going to spend some time this week to see what can be done and perhaps provide some timelines.

Feature request: https://s3drive.canny.io/feature-requests/p/support-self-signed-corporate-ssl-certificate
Thanks.

**Comment by HarrPerson** [comment]
Yes, it would be great if I could test the latest implementation. Where can I get the release? I have already installed the latest version (1.8.6) from the play store. With this version, I receive the same error message.

**Comment by tomekit** [comment]
Unfortunately there are still couple obstacles that we've found. 

We aim to support self-signed certificates, question is how long will it take for us to reliably support them on all platforms.

Let's keep this open, I will try to provide more updates soon.

**Comment by tomekit** [comment]
We've spent some time on it and came to conclusion that there isn't any feasible way that we could load the certificate added by the user on the OS level.

... but the good news is that we've built this feature in-app, so you can load your PEM certificate and make it trusted. It's part of the 1.8.7 release: https://s3drive.app/changelog which is released on all platforms except Android (awaiting Google approval ~2 days).

Please find the relevant docs: https://docs.s3drive.app/setup/self-signed-ssl/

I would be keen to hear if you've managed to import certificate into the app.

**Comment by tomekit** [comment]
Hi @HarrPerson,

I was wondering if you had chance to try out self-signed certificate in a newest Android release?

**Comment by HarrPerson** [comment]
I have updated the Android app to version _1.8.7_ but still receive the same error message.

I'm not sure if the documentation https://docs.s3drive.app/setup/self-signed-ssl/ is also relevant for the android app. But in Android I was not able to find the settings for importing the CA directly into the app. Only importing into the _android user certificate store_ is possible.

I also tried the latest AppImage (v1.8.7 (build: 10080701)) under Linux, which works fine also **without** importing the CA directly within the S3Driver application. The CA was imported in Linux before  in _"/etc/ca-certificates/trust-source/anchors/"_ and _"/etc/ssl/certs/"_.

**Comment by tomekit** [comment]
![Screenshot_20240504_211505](https://github.com/s3drive/app/assets/7759113/61eb56db-46a7-461e-bd9c-304e63320e76)
![Screenshot_20240504_211426](https://github.com/s3drive/app/assets/7759113/f5bfba84-698f-4f23-a63a-bfa6f04656e4)

**Comment by tomekit** [comment]
Thanks for letting me know.

Depending on the OS, system-level certificates might be used by S3Drive, but that was not the case for Android and the app level functionality was required.

On Android 1.8.7 you shall be able to see the import certificate option as per attached screenshots:
<img width="250" src="https://github.com/s3drive/app/assets/7759113/61eb56db-46a7-461e-bd9c-304e63320e76"/>
<img width="250" src="https://github.com/s3drive/app/assets/7759113/f5bfba84-698f-4f23-a63a-bfa6f04656e4"/>

Is "Trust SSL certificate (PEM)" option available for you on Android?

**Comment by HarrPerson** [comment]
How can I access the settings? There is no settings option displayed in my version:

![Screenshot_20240506-174304](https://github.com/s3drive/app/assets/11521502/fb6c450c-7095-48ef-91c7-1df651c62d77)

Also the blue button in the bottom right corner does not show any settings.

**Comment by tomekit** [comment]
We're awaiting major work on settings regrouping, so can make clear split between bucket specific settings (requiring login first) and global settings, which will subsequently allow us to expose app global settings to the initial screen. 

Since currently all settings (requiring login and global) are combined one place, unfortunately in order to access them you would need to configure a bucket with a valid SSL first (or you can create free S3Drive's email/password account to login to the app). 

From that point you shall be able to visit Settings. 
Bear in mind that as such SSL cert option won't be listed in web version, as web browsers have their own SSL trust mechanism.

**Comment by HarrPerson** [comment]
With an account I was able to add my own CA. With the imported own CA it worked also with my self-signed certificate. Thank you very much for the fast implementation and response.

I'm looking forward to the implementation of the settings improvements.

**Comment by tomekit** [comment]
Thanks, I am going to close this one for now.

There is an issue where by importing user generated CA all trusted ones stop working (for Rclone back-ends, but also Sync and Android mount): https://s3drive.canny.io/feature-requests/p/fix-user-x509-ca-breaking-trusted-certs-for-rclone-back-ends 
We will be addressing this in the future.
